File: legacy-ts/src/tools/task.ts
/**
 * Working memory task tools — Phase 5.
 *
 * task_start  — create task, seed context from brain Wiki + atoms
 * task_update — append findings, steps, artifacts, questions
 * task_complete — promote medium/high findings to Raw/curated/ queue, then clear
 * task_get    — read current task state or list active tasks
 */
import { z } from 'zod';
import type { CallToolResult } from '@modelcontextprotocol/sdk/types.js';
import {
  generateTaskId,
  createTask,
  getTask,
  getTaskState,
  updateTaskMeta,
  deleteTask,
  addStep,
  updateStep,
  addFinding,
  addArtifact,
  addQuestion,
  resolveQuestion,
  listActiveTasks,
  writeSnapshot,
  type StepStatus,
} from '../working/db.js';
import { seedTaskFromVault } from '../working/retrieval.js';
import { promoteTaskToVault } from '../working/promotion.js';
import { logger } from '../shared/logger.js';
import { formatError } from '../shared/errors.js';

// --- Zod schemas ---

const TaskStartSchema = z.object({
  goal: z.string().min(1).max(500),
  constraints: z.array(z.string().max(200)).max(20).optional(),
  plan: z.array(z.string().max(200)).max(50).optional(),
});

const TaskUpdateSchema = z.object({
  task_id: z.string().min(1),
  current_step: z.string().max(200).optional(),
  add_finding: z.object({
    content: z.string().min(1),
    importance: z.enum(['low', 'medium', 'high']).optional(),
    memory_type: z.enum(['semantic', 'episodic', 'procedural']).optional(),
  }).optional(),
  add_step: z.string().max(200).optional(),
  complete_step: z.number().int().positive().optional(),
  fail_step: z.number().int().positive().optional(),
  add_artifact: z.object({
    name: z.string().min(1).max(200),
    reference: z.string().min(1),
  }).optional(),
  add_question: z.string().max(500).optional(),
  resolve_question: z.object({
    id: z.number().int().positive(),
    resolution: z.string().min(1),
  }).optional(),
});

const TaskCompleteSchema = z.object({
  task_id: z.string().min(1),
  final_finding: z.string().max(2000).optional(),
});

const TaskGetSchema = z.object({
  task_id: z.string().optional(),
});

// --- Tool definitions ---

export const taskStartToolDefinition = {
  name: 'task_start',
  description:
    'Start a new working-memory task. Creates an in-session SQLite record, seeds context from brain Wiki pages and memory atoms, and returns a task_id. Use task_update to log findings and task_complete to promote them to the Wiki.',
  inputSchema: {
    type: 'object' as const,
    properties: {
      goal: { type: 'string', description: 'What this task is trying to accomplish' },
      constraints: { type: 'array', items: { type: 'string' }, description: 'Optional constraints or guardrails' },
      plan: { type: 'array', items: { type: 'string' }, description: 'Optional initial plan steps' },
    },
    required: ['goal'],
  },
};

export const taskUpdateToolDefinition = {
  name: 'task_update',
  description: 'Update working-memory task state. Append findings, add/complete steps, record artifacts, or track open questions.',
  inputSchema: {
    type: 'object' as const,
    properties: {
      task_id: { type: 'string', description: 'Task ID returned by task_start' },
      current_step: { type: 'string', description: 'Description of what is happening right now' },
      add_finding: {
        type: 'object',
        properties: {
          content: { type: 'string' },
          importance: { type: 'string', enum: ['low', 'medium', 'high'] },
          memory_type: { type: 'string', enum: ['semantic', 'episodic', 'procedural'] },
        },
        required: ['content'],
      },
      add_step: { type: 'string', description: 'Add a new pending step' },
      complete_step: { type: 'number', description: 'Mark a step ID as completed' },
      fail_step: { type: 'number', description: 'Mark a step ID as failed' },
      add_artifact: {
        type: 'object',
        properties: { name: { type: 'string' }, reference: { type: 'string' } },
        required: ['name', 'reference'],
      },
      add_question: { type: 'string', description: 'Record an open question' },
      resolve_question: {
        type: 'object',
        properties: { id: { type: 'number' }, resolution: { type: 'string' } },
        required: ['id', 'resolution'],
      },
    },
    required: ['task_id'],
  },
};

export const taskCompleteToolDefinition = {
  name: 'task_complete',
  description:
    'Mark a task as complete. Promotes medium/high findings to Raw/curated/ and queues them for brain_synthesize, then clears the task from working memory. Call brain_synthesize afterwards to write findings into the Wiki.',
  inputSchema: {
    type: 'object' as const,
    properties: {
      task_id: { type: 'string', description: 'Task ID to complete' },
      final_finding: { type: 'string', description: 'Optional summary finding to add before promoting (importance: high, type: episodic)' },
    },
    required: ['task_id'],
  },
};

export const taskGetToolDefinition = {
  name: 'task_get',
  description: 'Get the current state of a working-memory task, or list all active tasks if no task_id is provided.',
  inputSchema: {
    type: 'object' as const,
    properties: {
      task_id: { type: 'string', description: 'Task ID to retrieve (omit to list all active tasks)' },
    },
  },
};

// --- Handlers ---

export async function handleTaskStart(args: unknown): Promise<CallToolResult> {
  try {
    const input = TaskStartSchema.parse(args);
    const task_id = generateTaskId();
    createTask(task_id, input.goal, input.constraints, input.plan);

    if (input.plan) {
      for (const step of input.plan) addStep(task_id, step);
    }

    const { seeded, orphanedWarning } = await seedTaskFromVault(task_id, input.goal);
    logger.info('Task started', { task_id, goal: input.goal, seeded });
    writeSnapshot();

    return {
      content: [{
        type: 'text',
        text: [
          `Task started: ${task_id}`,
          `Goal: ${input.goal}`,
          input.constraints?.length ? `Constraints: ${input.constraints.join(', ')}` : null,
          input.plan?.length ? `Plan: ${input.plan.length} steps` : null,
          seeded > 0 ? `Seeded ${seeded} relevant results from brain (atoms + wiki)` : 'No matching brain context found',
          orphanedWarning ?? null,
        ].filter(Boolean).join('\n'),
      }],
    };
  } catch (err) {
    logger.error('task_start failed', { error: String(err) });
    return { content: [{ type: 'text', text: `Error: ${formatError(err)}` }], isError: true };
  }
}

export async function handleTaskUpdate(args: unknown): Promise<CallToolResult> {
  try {
    const input = TaskUpdateSchema.parse(args);
    const task = getTask(input.task_id);
    if (!task) return { content: [{ type: 'text', text: `Task not found: ${input.task_id}` }], isError: true };

    const changes: string[] = [];

    if (input.current_step !== undefined) {
      updateTaskMeta(input.task_id, { current_step: input.current_step });
      changes.push(`Current step: ${input.current_step}`);
    }
    if (input.add_finding) {
      addFinding(input.task_id, input.add_finding.content, input.add_finding.importance, input.add_finding.memory_type);
      changes.push(`Finding recorded (${input.add_finding.importance ?? 'medium'}, ${input.add_finding.memory_type ?? 'episodic'})`);
    }
    if (input.add_step) {
      const id = addStep(input.task_id, input.add_step);
      changes.push(`Step #${id} added`);
    }
    if (input.complete_step !== undefined) {
      updateStep(input.complete_step, 'completed' as StepStatus);
      changes.push(`Step #${input.complete_step} completed`);
    }
    if (input.fail_step !== undefined) {
      updateStep(input.fail_step, 'failed' as StepStatus);
      changes.push(`Step #${input.fail_step} marked failed`);
    }
    if (input.add_artifact) {
      const id = addArtifact(input.task_id, input.add_artifact.name, input.add_artifact.reference);
      changes.push(`Artifact #${id}: ${input.add_artifact.name}`);
    }
    if (input.add_question) {
      const id = addQuestion(input.task_id, input.add_question);
      changes.push(`Question #${id} added`);
    }
    if (input.resolve_question) {
      resolveQuestion(input.resolve_question.id, input.resolve_question.resolution);
      changes.push(`Question #${input.resolve_question.id} resolved`);
    }

    if (changes.length > 0) writeSnapshot();

    return {
      content: [{
        type: 'text',
        text: changes.length > 0
          ? `Task ${input.task_id} updated:\n${changes.join('\n')}`
          : `Task ${input.task_id}: no changes applied`,
      }],
    };
  } catch (err) {
    logger.error('task_update failed', { error: String(err) });
    return { content: [{ type: 'text', text: `Error: ${formatError(err)}` }], isError: true };
  }
}

export async function handleTaskComplete(args: unknown): Promise<CallToolResult> {
  try {
    const input = TaskCompleteSchema.parse(args);
    const state = getTaskState(input.task_id);
    if (!state) return { content: [{ type: 'text', text: `Task not found: ${input.task_id}` }], isError: true };

    if (input.final_finding) {
      addFinding(input.task_id, input.final_finding, 'high', 'episodic');
      const updated = getTaskState(input.task_id);
      if (updated) Object.assign(state, updated);
    }

    updateTaskMeta(input.task_id, { status: 'completed' });
    const counts = await promoteTaskToVault(state);
    deleteTask(input.task_id);
    writeSnapshot();

    logger.info('Task completed', { task_id: input.task_id, ...counts });

    return {
      content: [{
        type: 'text',
        text: [
          `Task ${input.task_id} completed.`,
          `Goal: ${state.task.goal}`,
          counts.created > 0
            ? `Promoted ${counts.created} finding(s) to Raw/curated/ — call brain_synthesize to write to Wiki.`
            : counts.skipped > 0 ? 'Findings already in vault (SHA256 match).' : 'No promotable findings.',
          'Working memory cleared.',
        ].join('\n'),
      }],
    };
  } catch (err) {
    logger.error('task_complete failed', { error: String(err) });
    return { content: [{ type: 'text', text: `Error: ${formatError(err)}` }], isError: true };
  }
}

export async function handleTaskGet(args: unknown): Promise<CallToolResult> {
  try {
    const input = TaskGetSchema.parse(args);

    if (!input.task_id) {
      const tasks = listActiveTasks();
      if (tasks.length === 0) return { content: [{ type: 'text', text: 'No active tasks.' }] };
      const summary = tasks.map(t => `- ${t.task_id}: ${t.goal} (started ${t.created_at})`).join('\n');
      return { content: [{ type: 'text', text: `Active tasks:\n${summary}` }] };
    }

    const state = getTaskState(input.task_id);
    if (!state) return { content: [{ type: 'text', text: `Task not found: ${input.task_id}` }], isError: true };

    const pending = state.steps.filter(s => s.status === 'pending').length;
    const done = state.steps.filter(s => s.status === 'completed').length;
    const openQs = state.questions.filter(q => !q.resolved).length;

    const lines = [
      `Task: ${state.task.task_id}`,
      `Goal: ${state.task.goal}`,
      `Status: ${state.task.status}`,
      state.task.current_step ? `Current step: ${state.task.current_step}` : null,
      `Steps: ${done} completed, ${pending} pending`,
      `Findings: ${state.findings.length}`,
      `Artifacts: ${state.artifacts.length}`,
      `Open questions: ${openQs}`,
    ];

    if (state.findings.length > 0) {
      lines.push('\nFindings:');
      for (const f of state.findings) {
        lines.push(`  [${f.importance}/${f.memory_type ?? 'episodic'}] ${f.content.slice(0, 120)}`);
      }
    }

    return { content: [{ type: 'text', text: lines.filter(Boolean).join('\n') }] };
  } catch (err) {
    logger.error('task_get failed', { error: String(err) });
    return { content: [{ type: 'text', text: `Error: ${formatError(err)}` }], isError: true };
  }
}

File: internal/mcp/task.go
package mcp

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/neverprepared/mcp-phantom-brain/internal/working"
)

// ---- task_start ----

func taskStartTool() mcp.Tool {
	return mcp.NewTool("task_start",
		mcp.WithDescription(
			`Open a new working-memory task. Returns a task_id you'll pass to `+
				`task_update / task_complete / task_get for the rest of the session. `+
				`Use at the top of any non-trivial multi-step work so findings and `+
				`artifacts have somewhere to land.`,
		),
		mcp.WithString("goal",
			mcp.Required(),
			mcp.Description("One-paragraph statement of what you're trying to accomplish."),
		),
		mcp.WithString("constraints",
			mcp.Description("Optional: constraints / non-goals / things that must not happen."),
		),
		mcp.WithString("plan",
			mcp.Description("Optional: initial plan or step outline. Free text; not parsed."),
		),
	)
}

func (s *Server) handleTaskStart(_ context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
	goal, err := req.RequireString("goal")
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}
	if strings.TrimSpace(goal) == "" {
		return mcp.NewToolResultError("goal must be non-empty"), nil
	}
	constraints, _ := req.RequireString("constraints")
	plan, _ := req.RequireString("plan")

	id, err := newTaskID()
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("alloc task_id: %v", err)), nil
	}
	if err := s.deps.Working.CreateTask(id, goal, constraints, plan); err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("create task: %v", err)), nil
	}
	return mcp.NewToolResultText(fmt.Sprintf("task_id: %s\n\nGoal: %s", id, goal)), nil
}

// newTaskID generates a 16-hex-char task id. Random enough for one
// session's worth of tasks; not a UUID — we don't need universal
// uniqueness, just per-process uniqueness within wm-<PID>.sqlite's
// PRIMARY KEY constraint.
func newTaskID() (string, error) {
	var b [8]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", err
	}
	return hex.EncodeToString(b[:]), nil
}

// ---- task_update ----

func taskUpdateTool() mcp.Tool {
	return mcp.NewTool("task_update",
		mcp.WithDescription(
			`Append progress to an open task. Use 'type' to discriminate: `+
				`'finding' (logs an observation), 'artifact' (records a file/PR/URL), `+
				`'question' (logs something unresolved), 'step' (appends a planned step), `+
				`or 'current_step' (sets the in-progress step description). `+
				`Medium and high importance findings are promoted to Raw/curated/ at task_complete.`,
		),
		mcp.WithString("task_id",
			mcp.Required(),
			mcp.Description("The id returned by task_start."),
		),
		mcp.WithString("type",
			mcp.Required(),
			mcp.Description("One of: finding | artifact | question | step | current_step."),
		),
		mcp.WithString("content",
			mcp.Description("For finding/question/step/current_step: the prose body. Required for those types."),
		),
		mcp.WithString("importance",
			mcp.Description("For type=finding: low | medium | high. Default medium. Only medium and high are promoted at task_complete."),
		),
		mcp.WithString("memory_type",
			mcp.Description("For type=finding: semantic | episodic | procedural. Optional but useful for downstream synthesis."),
		),
		mcp.WithString("name",
			mcp.Description("For type=artifact: human-readable name (e.g. 'API spec PDF', 'PR #42')."),
		),
		mcp.WithString("reference",
			mcp.Description("For type=artifact: the actual file path / URL / PR link."),
		),
	)
}

func (s *Server) handleTaskUpdate(_ context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
	taskID, err := req.RequireString("task_id")
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}
	updateType, err := req.RequireString("type")
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	if _, err := s.deps.Working.GetTask(taskID); err != nil {
		if errors.Is(err, working.ErrTaskNotFound) {
			return mcp.NewToolResultError(fmt.Sprintf("no task with id %q (call task_start first)", taskID)), nil
		}
		return mcp.NewToolResultError(fmt.Sprintf("get task: %v", err)), nil
	}

	switch strings.ToLower(strings.TrimSpace(updateType)) {
	case "finding":
		return s.taskUpdateFinding(req, taskID)
	case "artifact":
		return s.taskUpdateArtifact(req, taskID)
	case "question":
		return s.taskUpdateQuestion(req, taskID)
	case "step":
		return s.taskUpdateStep(req, taskID)
	case "current_step":
		return s.taskUpdateCurrentStep(req, taskID)
	default:
		return mcp.NewToolResultError(fmt.Sprintf(
			"unknown type %q (want finding|artifact|question|step|current_step)", updateType)), nil
	}
}

func (s *Server) taskUpdateFinding(req mcp.CallToolRequest, taskID string) (*mcp.CallToolResult, error) {
	content, _ := req.RequireString("content")
	if strings.TrimSpace(content) == "" {
		return mcp.NewToolResultError("finding requires content"), nil
	}
	importance := strings.TrimSpace(strings.ToLower(orDefault(req, "importance", working.ImportanceMedium)))
	memoryType := strings.TrimSpace(strings.ToLower(orDefault(req, "memory_type", "")))

	id, err := s.deps.Working.AppendFinding(taskID, content, importance, memoryType)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("append finding: %v", err)), nil
	}
	return mcp.NewToolResultText(fmt.Sprintf("Finding #%d logged (%s).", id, importance)), nil
}

func (s *Server) taskUpdateArtifact(req mcp.CallToolRequest, taskID string) (*mcp.CallToolResult, error) {
	name, _ := req.RequireString("name")
	reference, _ := req.RequireString("reference")
	if strings.TrimSpace(name) == "" || strings.TrimSpace(reference) == "" {
		return mcp.NewToolResultError("artifact requires name + reference"), nil
	}
	id, err := s.deps.Working.AppendArtifact(taskID, name, reference)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("append artifact: %v", err)), nil
	}
	return mcp.NewToolResultText(fmt.Sprintf("Artifact #%d logged: %s -> %s", id, name, reference)), nil
}

func (s *Server) taskUpdateQuestion(req mcp.CallToolRequest, taskID string) (*mcp.CallToolResult, error) {
	content, _ := req.RequireString("content")
	if strings.TrimSpace(content) == "" {
		return mcp.NewToolResultError("question requires content"), nil
	}
	id, err := s.deps.Working.AppendQuestion(taskID, content)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("append question: %v", err)), nil
	}
	return mcp.NewToolResultText(fmt.Sprintf("Question #%d logged.", id)), nil
}

func (s *Server) taskUpdateStep(req mcp.CallToolRequest, taskID string) (*mcp.CallToolResult, error) {
	content, _ := req.RequireString("content")
	if strings.TrimSpace(content) == "" {
		return mcp.NewToolResultError("step requires content"), nil
	}
	id, err := s.deps.Working.AppendStep(taskID, content)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("append step: %v", err)), nil
	}
	return mcp.NewToolResultText(fmt.Sprintf("Step #%d appended.", id)), nil
}

func (s *Server) taskUpdateCurrentStep(req mcp.CallToolRequest, taskID string) (*mcp.CallToolResult, error) {
	content, _ := req.RequireString("content")
	if strings.TrimSpace(content) == "" {
		return mcp.NewToolResultError("current_step requires content"), nil
	}
	if err := s.deps.Working.SetCurrentStep(taskID, content); err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("set current step: %v", err)), nil
	}
	return mcp.NewToolResultText(fmt.Sprintf("Current step set: %s", content)), nil
}

// orDefault returns the trimmed string at key, or def if missing/empty.
func orDefault(req mcp.CallToolRequest, key, def string) string {
	got, _ := req.RequireString(key)
	if strings.TrimSpace(got) == "" {
		return def
	}
	return got
}

// ---- task_complete ----

func taskCompleteTool() mcp.Tool {
	return mcp.NewTool("task_complete",
		mcp.WithDescription(
			`Mark a task complete and promote its medium/high importance findings to `+
				`Raw/curated/ as a single aggregated note. After this, the task is `+
				`searchable via brain_recall like any other curated knowledge. The task `+
				`row stays in working memory (queryable via task_get) until the orphan `+
				`reaper sweeps the process's shard at next startup.`,
		),
		mcp.WithString("task_id",
			mcp.Required(),
			mcp.Description("The task to complete."),
		),
	)
}

func (s *Server) handleTaskComplete(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
	taskID, err := req.RequireString("task_id")
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}
	task, err := s.deps.Working.GetTask(taskID)
	if err != nil {
		if errors.Is(err, working.ErrTaskNotFound) {
			return mcp.NewToolResultError(fmt.Sprintf("no task with id %q", taskID)), nil
		}
		return mcp.NewToolResultError(fmt.Sprintf("get task: %v", err)), nil
	}

	important, err := s.deps.Working.ListImportantFindings(taskID)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("list findings: %v", err)), nil
	}
	artifacts, err := s.deps.Working.ListArtifacts(taskID)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("list artifacts: %v", err)), nil
	}
	questions, err := s.deps.Working.ListQuestions(taskID)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("list questions: %v", err)), nil
	}

	// If nothing was logged, complete the task without writing a
	// promoted note — there'd be nothing to index.
	if len(important) == 0 && len(artifacts) == 0 && len(questions) == 0 {
		if err := s.deps.Working.CompleteTask(taskID); err != nil {
			return mcp.NewToolResultError(fmt.Sprintf("mark complete: %v", err)), nil
		}
		return mcp.NewToolResultText(fmt.Sprintf("Task %s marked complete (no findings to promote).", taskID)), nil
	}

	body := renderPromotedTaskBody(task, important, artifacts, questions)
	title := fmt.Sprintf("Task: %s", strings.TrimSpace(task.Goal))

	res, errMsg, ok := s.ingestMarkdown(ctx, ingestParams{
		Subdir:    "curated",
		StampKey:  "completed_at",
		Content:   body,
		Title:     title,
		Filename:  fmt.Sprintf("task-%s.md", taskID),
		SourceURL: "",
	})
	if !ok {
		return mcp.NewToolResultError(fmt.Sprintf("promote: %s", errMsg)), nil
	}

	if err := s.deps.Working.CompleteTask(taskID); err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("mark complete: %v", err)), nil
	}

	status := res.Status
	if status == "stored" {
		return mcp.NewToolResultText(fmt.Sprintf(
			"Task %s complete. Promoted %d finding(s), %d artifact(s), %d question(s) to %s. SHA: %s.",
			taskID, len(important), len(artifacts), len(questions), res.RelativePath, res.SHA,
		)), nil
	}
	return mcp.NewToolResultText(fmt.Sprintf(
		"Task %s complete. Promotion was a duplicate of an existing note (SHA: %s).", taskID, res.SHA,
	)), nil
}

func renderPromotedTaskBody(task *working.Task, findings []working.Finding, artifacts []working.Artifact, questions []working.Question) string {
	var b strings.Builder

	if task.Goal != "" {
		fmt.Fprintf(&b, "**Goal:** %s\n\n", strings.TrimSpace(task.Goal))
	}
	if task.Constraints != "" {
		fmt.Fprintf(&b, "**Constraints:** %s\n\n", strings.TrimSpace(task.Constraints))
	}
	if task.Plan != "" {
		fmt.Fprintf(&b, "**Plan:** %s\n\n", strings.TrimSpace(task.Plan))
	}

	if len(findings) > 0 {
		b.WriteString("## Findings\n\n")
		for _, f := range findings {
			fmt.Fprintf(&b, "### [%s] ", strings.ToUpper(f.Importance))
			if f.MemoryType != "" {
				fmt.Fprintf(&b, "(%s) ", f.MemoryType)
			}
			fmt.Fprintf(&b, "#%d — %s\n\n", f.ID, f.CreatedAt.Format(time.RFC3339))
			b.WriteString(strings.TrimSpace(f.Content))
			b.WriteString("\n\n")
		}
	}
	if len(artifacts) > 0 {
		b.WriteString("## Artifacts\n\n")
		for _, a := range artifacts {
			fmt.Fprintf(&b, "- **%s** — %s\n", a.Name, a.Reference)
		}
		b.WriteString("\n")
	}
	if len(questions) > 0 {
		b.WriteString("## Open Questions\n\n")
		for _, q := range questions {
			marker := "?"
			if q.Resolved {
				marker = "+"
			}
			fmt.Fprintf(&b, "- (%s) %s\n", marker, q.Question)
			if q.Resolved && q.Resolution != "" {
				fmt.Fprintf(&b, "   - **Resolution:** %s\n", q.Resolution)
			}
		}
		b.WriteString("\n")
	}
	return b.String()
}

// ---- task_get ----

func taskGetTool() mcp.Tool {
	return mcp.NewTool("task_get",
		mcp.WithDescription(
			`Read full state for a task: goal, status, current step, all findings, `+
				`artifacts, and questions. Returns whether the task is still active.`,
		),
		mcp.WithString("task_id",
			mcp.Required(),
			mcp.Description("The task to inspect."),
		),
	)
}

func (s *Server) handleTaskGet(_ context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
	taskID, err := req.RequireString("task_id")
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}
	task, err := s.deps.Working.GetTask(taskID)
	if err != nil {
		if errors.Is(err, working.ErrTaskNotFound) {
			return mcp.NewToolResultError(fmt.Sprintf("no task with id %q", taskID)), nil
		}
		return mcp.NewToolResultError(fmt.Sprintf("get task: %v", err)), nil
	}
	findings, _ := s.deps.Working.ListFindings(taskID)
	artifacts, _ := s.deps.Working.ListArtifacts(taskID)
	questions, _ := s.deps.Working.ListQuestions(taskID)
	steps, _ := s.deps.Working.ListSteps(taskID)

	var b strings.Builder
	fmt.Fprintf(&b, "# Task %s — %s\n\n", task.ID, task.Status)
	fmt.Fprintf(&b, "**Goal:** %s\n\n", task.Goal)
	if task.Constraints != "" {
		fmt.Fprintf(&b, "**Constraints:** %s\n\n", task.Constraints)
	}
	if task.Plan != "" {
		fmt.Fprintf(&b, "**Plan:** %s\n\n", task.Plan)
	}
	if task.CurrentStep != "" {
		fmt.Fprintf(&b, "**Current step:** %s\n\n", task.CurrentStep)
	}
	fmt.Fprintf(&b, "Created: %s · Updated: %s\n\n", task.CreatedAt.Format(time.RFC3339), task.UpdatedAt.Format(time.RFC3339))

	if len(steps) > 0 {
		b.WriteString("## Steps\n\n")
		for _, st := range steps {
			marker := "[ ]"
			if st.Status == working.StepDone {
				marker = "[x]"
			}
			fmt.Fprintf(&b, "- %s #%d %s\n", marker, st.ID, st.Description)
		}
		b.WriteString("\n")
	}
	if len(findings) > 0 {
		b.WriteString("## Findings\n\n")
		for _, f := range findings {
			fmt.Fprintf(&b, "- [%s", f.Importance)
			if f.MemoryType != "" {
				fmt.Fprintf(&b, "/%s", f.MemoryType)
			}
			fmt.Fprintf(&b, "] #%d — %s\n", f.ID, strings.TrimSpace(f.Content))
		}
		b.WriteString("\n")
	}
	if len(artifacts) > 0 {
		b.WriteString("## Artifacts\n\n")
		for _, a := range artifacts {
			fmt.Fprintf(&b, "- #%d **%s** — %s\n", a.ID, a.Name, a.Reference)
		}
		b.WriteString("\n")
	}
	if len(questions) > 0 {
		b.WriteString("## Questions\n\n")
		for _, q := range questions {
			marker := "open"
			if q.Resolved {
				marker = "resolved"
			}
			fmt.Fprintf(&b, "- (%s) #%d %s\n", marker, q.ID, q.Question)
		}
		b.WriteString("\n")
	}
	return mcp.NewToolResultText(b.String()), nil
}
